File: BST.py
#BST implementation using python

class Node:
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None

#insertion of a node
def insert(root,data):
    temp = Node(data)

    if root==None:
        return temp
    
    elif data<=root.data:
        root.left = insert(root.left,data)

    elif data>root.data:
        root.right = insert(root.right,data)

    return root

#inorder traversal
def display(root):
    if root==None:
        return None
    
    display(root.left)
    print(root.data,end = " ")
    display(root.right)

#deletion of a node
def deletion(root,value):

    if root == None:
        return None
    
    if root.data<value:
        root.right = deletion(root.right,value)
    elif root.data>value:
        root.left = deletion(root.left,value)
    
    elif root.data == value:

        if root.left==None and root.right==None:
            root = None
            return root
        
        elif root.left!=None and root.right==None:
            root = root.left
        
        elif root.right!=None and root.left==None:
            root = root.right

        else:
            temp = root.left
            while temp.right!=None:
                temp = temp.right
            root.data = temp.data
            root.left = deletion(root.left,root.data)

    return root

#main

root = Node(int(input("Enter the root node --> ")))

ch = 1
while ch == 1:
    print("To enter more nodes, press 1\nTo exit press 0")
    ch = int(input("\nEnter your choice : "))

    if ch!=1:
        break
    root = insert(root,int(input("Enter the value of element --> ")))
    

File: BST.c
#include <stdio.h>
#include <stdlib.h>

//Structure of a BINARY SEARCH TREE
struct Node {
    int data;
    struct Node * left;
    struct Node * right;
};

struct Node * insert(struct Node * root, int value){
    if(root==NULL){
        struct Node * temp = (struct Node *)malloc(sizeof(struct Node));
        temp->data = value;
        return temp;
    }
    if(value<root->data){
        root->left = insert(root->left,value);
    }
    else if(value>=root->data){
        root->right = insert(root->right,value);
    }

    return root;
}

//Preorder Traversal
void inorder(struct Node * root){
    
    if(root==NULL){
        return;
    }

    inorder(root->left);
    printf("%d ",root->data);
    inorder(root->right);
    
}

void delete(struct Node * root, int value){
    int x;
    
    if(root==NULL){
        printf("Empty BST!");
        return;
    }

    if(root->data == value){

        if(root->left==NULL && root->right==NULL){
            printf("Node deleted with value = %d",value);
            free(root);
            return;
        }

        else if(root->left==NULL && root->right!=NULL){
            root = root->right;
            printf("Node deleted with value = %d",value);
            return;
        }

        else if(root->left!=NULL && root->right==NULL){
            root = root->left;
            printf("Node deleted with value = %d",value);
            return;
        }

        else{
            root->data = x;
            printf("Node deleted with value = %d",value);
        }

    }
    delete(root->left,value);
    delete(root->right,value);
    x = root->data;
}

int main(){
    struct Node * root = NULL;

    for(int i=1; i<=10; i++){
        root = insert(root,i);
    }
    inorder(root);

    delete(root,5);
    printf("\nTree after deletion :\n");
    inorder(root);
    
    return 0;
}
